Toggle button emits ButtonClick event

xaptronic opened the pull request:
This PR allows one to programmatically set the state of the Toggle button without invoking the click callback.

I did not find any existing tests for this, and seemed to have some issues running `node make test:unit` so did not end up that far down this path.

I spot checked a number of existing example uses of this button and there does not appear to be a change in function.

Looking forward to any feedback.

- [x] issues: fixes #10424 
- [x] tests added / passed
- [ ] release document entry (if new feature or API change)

**Summary of Buttons and their Use**
  | on_change | js_on_change | on_event | js_on_event | on_click | js_on_click
-- | -- | -- | -- | -- | -- | --
Button | no | no | yes | yes | deprecated | deprecated
Toggle | yes | yes | yes | yes | deprecated | deprecated
Dropdown | no | no | yes | yes | deprecated | deprecated
CheckboxGroup | yes | yes | no | no | removed | removed
RadioGroup | yes | yes | no | no | removed | removed
CheckboxButtonGroup | yes | yes | yes | yes | removed | removed
RadioButtonGroup | yes | yes | yes | yes | removed | removed

xaptronic commented:
Ah there we go, integration test failure - should we pass the toggle state into the callback from `on_event`?

bryevdv commented:
> should we pass the toggle state into the callback from on_event?

I'm ambivalent about that. It seems redundant since a click even callback can just inspect the `active` property on the button. We just need to make sure the event is triggered last to guarantee that the `active` state is guaranteed to be updated first. 

Also as suggested in the OP, I think wed can just drop `on_click` and `js_on_click` altogether. The 3.0 release is a perfect time for a change like that.

Regarding the test, the file that needs to be updated is`tests/integration/widgets/test_toggle.py` Th test should be changed to us `on_event` and `js_on_event` with a `ButtonClick` event.

xaptronic commented:
> will need a note in the migration guide of the release notes

Removed `on_click` and `js_on_click`, updated the integration test and also changed the examples in the documentation.

Where would I find the migration guide, or are the release notes something done when the new version is cut?

bryevdv commented:
Hi @xaptronic Sorry I probably should have been a bit more thorough in explanation: if we are going to remove `on_click` and `js_on_click` in favor of generic `on_event` and `js_on_event`, then we need to do it *everywhere*. That includes other button classes, as well as all tests an examples that might currently use it. A quick grep shows maybe ~50 instances that would need changing. There's not any particular hurry but you will have to let us know if you are up for that larger task. 

Also after some though, it would probably be OK to leave the definitions of of `on_click` and `js_on_click` in place, but add deprecations to them. All of the examples, tests, etc would need to be updated to use events, regardless.

xaptronic commented:
I'm interested in contributing.

I'm not 100% sold on removing the on_click/js_on_click, it seems to me like a useful a dev friendly shortcut for what I would imagine to be a common pattern - is there already a consensus on this change among other devs regarding this and moving towards using `on_event` with an event type going forward?

bryevdv commented:
@xaptronic Note you can also use the string alias for the event, sparing an import:
```
button.js_on_event('button_click', ...)
```
To me this is roughly equivalent to `js_on_click` but has the benefit of being the same mechanism as all other events, which is one thing documented in one consistent way. Happy to hear any other thoughts (cc @bokeh/dev) but it would take alot to convince me not to proceed with deprecating this in favor of consistency.

xaptronic commented:
This makes sense to me. I think I was under the impression that there were more types of event handlers, but now looking, it seems there aren't actually that many.

I will start to look at instances on `js/on_click` and what it will take to update them.

Thanks for the feedback.

bryevdv commented:
@xaptronic Great! we have a `deprecated` function you can call in all the `on_click` / `js_on_click` methods, just state that the are deprecated as of Bokeh 3.0 and state what to use instead (you can great the repo for some examples of typical deprecation messages)

xaptronic commented:
Do you have any preference if I continue to work with this PR, or should I start another?

xaptronic commented:
I have done a round of updating the `buttons` with deprecations on `on_click/js_on_click` and updated relevant tests, examples and sphinx (docs?) entries. Take a look and let me know if the approach is sound, and I will continue with the radio/check boxes.

I accidentally put the same commit message twice, one of them is the changes, the other was just merging updates from master - I read in the contrib guide that you'd discouraging people force pushing.. hopefully we can squash these when the time comes.

While testing some examples, it seems that this one here doesn't work unless a title is specified on the figure initially: https://github.com/bokeh/bokeh/blob/branch-3.0/examples/reference/models/button_server.py

Let me know what the best action would be to take with that upfront.

bryevdv commented:
> I read in the contrib guide that you'd discouraging people force pushing..

I think this is outdated (cc @tcmetzger) there's not reason for anyone to avoid force-pushing on their own fork. We have release branches protected so it's really not an issue. Feel free to rebase, etc as you see fit. 

> While testing some examples, it seems that this one here doesn't work unless a title is specified on the figure initially:

Seems like a bug, I am not aware of an issue offhand (cc @mattpap) so feel free to make a new one (maybe do a quick search first in case I am just forgetting)

bryevdv commented on bokeh/models/widgets/buttons.py:
```diff
@@ -121,24 +124,22 @@ class Toggle(AbstractButton):
 
     def on_click(self, handler: Callable[[bool], None]) -> None:
         """ Set up a handler for button state changes (clicks).
-
```
We use blank lines between summary/args/returns all throughout the codebase, so these should be restored just for consistency

xaptronic commented:
Started updating the checkbox/radio button variants - one question: with the `CheckboxGroup` and `RadioGroup`, it strikes as odd for them to emit `ButtonClick` events, but implemented this way, it does actually work - any opinions on how this is done?

Additionally, in the callback for `js_on_event`, `this` / `cb_obj` no longer refers to the source element, and so now properties of the control is available under `this.origin` / `cb_obj.origin`.

With `on_click`, the `event` object passed into the callback doesn't have any reference to the originating control, so for example, if one had multiple buttons that used the same callback, the callback would need to be wrapped in a lambda function in order to pass the event source in.

Curious how this sits with you?

tcmetzger commented:
> I read in the contrib guide that you'd discouraging people force pushing..

@xaptronic could you point me to where in the docs that get's mentioned? I'd be happy to get that ironed out!

xaptronic commented:
> @xaptronic could you point me to where in the docs that get's mentioned? I'd be happy to get that ironed out!

I will try to find it, it seems to escape me now. Something about not rebasing but merging updates from upstream.

xaptronic commented:
Here it is: https://github.com/bokeh/bokeh/blob/branch-3.0/.github/CONTRIBUTING.md#creating-a-pull-request-pr

>Do NOT rebase, amend, or cherry-pick published commits. Any of those actions will make us start the review from scratch. If you need updates from master, just merge it into your branch.

mattpap commented:
> While testing some examples, it seems that this one here doesn't work unless a title is specified on the figure initially:

It's a bug. I just fixed a similar issue with plot's toolbar in a recent PR.

bryevdv commented:
> Started updating the checkbox/radio button variants - one question: with the CheckboxGroup and RadioGroup, it strikes as odd for them to emit ButtonClick events, but implemented this way, it does actually work - any opinions on how this is done?

Not offhand, ultimately something on the BokehJS side must be emitting the button clicked event already. 

> With on_click, the event object passed into the callback doesn't have any reference to the originating control, so for example, if one had multiple buttons that used the same callback, the callback would need to be wrapped in a lambda function in order to pass the event source in.
> 
> Curious how this sits with you?

We just need to over-communicate any changes and/or deprecations. The goal was to be able to distinguish programmatic changes to `active` etc. from actual UI button clicks, so we have to live with this difference regardless.

xaptronic commented:
> Not offhand, ultimately something on the BokehJS side must be emitting the button clicked event already.

Right, something should be emitted; to ask more directly, are we ok with (at least for now) radio buttons and checkboxes emitting a click event called "button_click" - naming wise maybe it's not the best match, but it also seems like a waste of energy to add an event called something like "radio_click".

> so we have to live with this difference regardless
A possible solution to this could be that server side event objects contain the object/widget that was the source of the event (like is the case with js_on_event - it has `origin` property) - this would at least make the python/js side of this consistent.

xaptronic commented:
Another thought - what if we rename `ButtonClick` to just click? I could imagine many things might want to make use of a click event..

bryevdv commented:
> Another thought - what if we rename ButtonClick to just Click? I could imagine many things might want to make use of a click event..

That's a larger discussion, for this PR I think `button_click` is fine for now (and also does not seem lit too much of a stretch to me for radio buttons fwiw)

bryevdv commented on bokeh/models/widgets/buttons.py:
```diff
@@ -133,12 +135,13 @@ def on_click(self, handler: Callable[[bool], None]) -> None:
-
         """
+        deprecated((3, 0, 0), 'on_click(handler)', 'on_event("button_click", handler)')
         self.on_change('active', lambda attr, old, new: handler(new))
```
This (and elsewhere) should change to use the event, to be consistent with other widgets that emit button click events I think

xaptronic replied:
Wouldn't that change the behavior of a method that we are intending to be deprecated?

xaptronic commented on sphinx/source/docs/user_guide/examples/interaction_checkbox_button_group.py:
```diff
@@ -4,8 +4,8 @@
-checkbox_button_group.js_on_click(CustomJS(code="""
-    console.log('checkbox_button_group: active=' + this.active, this.toString())
+checkbox_button_group.js_on_event("button_click", CustomJS(code="""
+    console.log('checkbox_button_group: active=' + this.origin.active, this.toString())
```
Sounds good, will make that update in the examples/tests.

bryevdv replied:
It's probably not as clear in the docs as it needs to be but `cb_obj` means different things for property change (`js_on_change`) callbacks than it does for event (`js_on_event`) callbacks. For property change callbacks, it contains the object that has the property the callback was on ("callback object"). That was thecase for nearly a decade, long before event callbacks existed. When event callbacks were added, we wanted to be able to have users re-use the same `CustomJS` objects. That meant we had to pass _something_ as `cb_obj` to the JS code. But it does not make sense to pass the "callback object" and in fact sometimes there is not one at all, e.g. for `Document` callbacks. But there are "event objects" that contain event-specific information that callbacks might need. That is what is passed for `cb_obj` for event callbacks. 

You can see all of them here:

https://docs.bokeh.org/en/latest/docs/reference/events.html

Some of them (e.g. button click) are empty, but most of them are not. Button click is empty because there is no further information to pass along, the information is simply that a button click occurred and the event itself conveys that. But in any case, they are not useless in general, because in general they contain more information. However, this is also why for event callbacks, if references to models are needed, we (or at least I) recommend using the `args` dict. Also this really on affects the "stateful buttons".  Options could be to make new different "stateful button click" events but that starts to get overly complex for users IMO, or as @mattpap suggests, just not have  these be button click emitters at all. 

TLDR; historical quirks in an old OSS project

xaptronic replied:
I think we're good here.

mattpap commented on examples/models/file/buttons.py:
```diff
@@ -6,42 +6,43 @@
 
 checkbox_button_group = CheckboxButtonGroup(labels=["Option 1", "Option 2", "Option 3"], active=[0, 1])
-checkbox_button_group.js_on_click(CustomJS(code="console.log('checkbox_button_group: active=' + this.active, this.toString())"))
+checkbox_button_group.js_on_event("button_group", CustomJS(code="console.log('checkbox_button_group: active=' + this.origin.active, this.toString())"))
```
`button_group` event doesn't exist.

mattpap commented on bokehjs/src/lib/models/widgets/radio_group.ts:
```diff
@@ -21,7 +22,10 @@ export class RadioGroupView extends InputGroupView {
-      radio.addEventListener("change", () => this.change_active(i))
+      radio.addEventListener("click", () => {
+        this.change_active(i)
+        this.model.trigger_event(new ButtonClick())
```
Radios and checkboxes aren't buttons, just stateful widgets. Checking their `active` property change should be sufficient. Besides this, `click` is much more limited event that `change`, degrading functionality of this widget.

bryevdv replied:
They had `on_click` already for some reason, though. If you are OK just dropping those entirely at 3.0 (just Radios and Checkboxes) then this could be dispensed with.

xaptronic replied:
The original motivation behind this PR was to allow Toggle buttons to have their state set programatically without invoking the click callback - which if you had an `on_click` set on a Toggle, setting it's `active` state would cause the click callback to fire.

If we drop click events on this radio/checkbox group buttons, the above issue would remain with these widgets. I can buy the argument that radio/checkboxes are strictly stateful widgets, and a click event is not useful, however, I'd only point out that you can infact have click and change handlers on them in Javascript land.

I personally don't use radios/checkboxes very often so I'm not strongly opinionated either way.

bryevdv replied:
I guess the question is whether there is any demand to be able to differentiate real UI clicks of "stateful buttons" from programmatic assignment in the same way as "plain buttons". I don't have strong feelings of preferences about that. These options for "stateful buttons" would be ok to me:

* have them emit `ButtonClick` and have them keep a (deprecated) `on_click` that responds to the _event_ (to be consistent with every other `on_click` at this point)
* have them not emit `ButtonClick`, but remove `on_click` immediately

What I would be 👎  on would be to have some object emit `ButtonClick` but not have an `on_click` (or vice versa) or to have some `on_click` methods mean different things than other `on_click` methods.

xaptronic replied:
I'm inclined to just remove the `on_click` and revert the Javascript change here (option 2).

bryevdv replied:
That seems to be what @mattpap also desires, so 👍 from me to  proceed with that (just need a minimal note in the release notes about the changes: https://github.com/bokeh/bokeh/blob/branch-3.0/sphinx/source/docs/releases/3.0.0.rst) Thanks for your patience with some back and forth @xaptronic !

xaptronic replied:
>I guess I am + 0.5 to keep these other "stateful buttons" doing the same, for that reason.

Will proceed to remove the `on_click` from the radio/checkbox group buttons, but let's not forget about your comment here, as it could be something to revisit in the future. Perhaps there is some wisdom to having the events be similar to what you can do natively in the browser, or having some kind of system where you can freely attach any browser event to an event and listen on the python side - much larger discussion way outside the scope of these changes, and certainly more than enough on the queue for now :)

mattpap commented on bokehjs/src/lib/models/widgets/button_group.ts:
```diff
@@ -1,9 +1,10 @@
 import {div} from "core/dom"
 import * as p from "core/properties"
 
+
```
Empty line should be removed.

bryevdv commented on sphinx/source/docs/releases/3.0.0.rst:
```diff
@@ -148,6 +154,9 @@ Deprecations
 instead.
 
+``on_click`` event handlers on button type widgets have been replaced with ``on_event``
+using ``bokeh.events.ButtonClick`` as the event type.
```
```suggestion
using ``bokeh.events.ButtonClick`` ("button_click") as the event type.
```

bryevdv commented on bokehjs/src/lib/models/widgets/button_group.ts:
```diff
@@ -39,7 +39,10 @@ export abstract class ButtonGroupView extends OrientedControlView {
-      button.addEventListener("click", () => this.change_active(i))
+      button.addEventListener("click", () => {
+        this.change_active(i)
+        this.model.trigger_event(new ButtonClick())
```
We were ging to remove ButtonClick from stateful buttons? radio/checkbox group should use "on change" for both Python and JS

bryevdv commented:
@xaptronic thanks for you patience just one small comment

bryevdv commented:
If you have the bandwidth @xaptronic it would be really helpful to have all the current state summarized in a table. We could try to work that in to the docs somewhere later.

xaptronic commented:
> would be really helpful to have

Can you clarify what you mean by "current state"?

bryevdv commented:
> Can you clarify what you mean by "current state"?

I was thinking of a summary of widget type vs the various different methods of use available/appropriate for them (on_click, on_event, on_change, etc)

bryevdv commented:
@mattpap this LGTM I'll plan to merge tomorrow morning unless have any comments (or ask for more time)

bryevdv commented:
Thanks @xaptronic !
